device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -918,29 +935,27 @@ else if (sendResult == SendResult.SUBSCRIPTION_IN_PROGRESS)
             {
                 // Proton-j doesn't handle the scenario of sending twin/method messages on links that haven't been opened remotely yet, so hold
                 // off on sending them until the subscription has finished.
-                log.trace("Attempted to send twin/method message while the twin/method subscription was in progress. Adding it back to messages to send queue to try again after the subscription has finished ({})", message);
-                messagesToSend.add(message);
+                TransportException transportException = new TransportException("Amqp message failed to send because it is a twin/method message that can't be sent while the twin/method subscription is in progress. " + message);
+                transportException.setRetryable(true);
+                log.debug("Notifying transport layer that amqp message failed to send.", transportException);
+                this.listener.onMessageSent(message, message.getConnectionDeviceId(), transportException);
             }
             else if (sendResult == SendResult.LINKS_NOT_OPEN)
             {
                 // Shouldn't happen. If it does, it signals that we have a bug in this SDK.
-                log.warn("Failed to send a message because its AMQP links were not open yet. Adding it back to messages to send queue ({})", message);
-                messagesToSend.add(message);
+                TransportException transportException = new TransportException("Amqp message failed to send because its AMQP links were not open yet. " + message);
+                transportException.setRetryable(true);
+                log.debug("Notifying transport layer that amqp message failed to send.", transportException);
+                this.listener.onMessageSent(message, message.getConnectionDeviceId(), transportException);
             }
             else if (sendResult == SendResult.UNKNOWN_FAILURE)
             {
                 // Shouldn't happen. If it does, it signals that we have a bug in this SDK.
-                log.warn("Unknown failure occurred while attempting to send. Adding it back to messages to send queue ({})", message);
-                messagesToSend.add(message);
+                TransportException transportException = new TransportException("Amqp message failed to send for unknown reasons. " + message);
+                transportException.setRetryable(true);
+                log.debug("Notifying transport layer that amqp message failed to send.", transportException);
+                this.listener.onMessageSent(message, message.getConnectionDeviceId(), transportException);
             }
-
-            message = messagesToSend.poll();
-        }
-
-        if (message != null)
```

Comment:
instead of doing
```
message = queue.poll();
while (message != null)
{
  // send message
  message = queue.poll();
}
queue.enqueue(message);
```

we'll do

```
while (queue.peek() != null)
{
  message = queue.poll();
  // send message
}
```

for simplicity and so that the ordering of messages isn't unnecessarily jumbled